src/main/java/com/minecolonies/coremod/commands/AbstractCommandParser.java

```diff
@@ -0,0 +1,678 @@
+    {
+        final Map<String, TreeNode<IMenu>> map = new HashMap<>();
+        final List<TreeNode<IMenu>> children = parentTreeNode.getChildren();
+        for (final TreeNode<IMenu> treeNode : children)
+        {
+            map.put(treeNode.getData().getMenuItemName().toLowerCase(Locale.ROOT), treeNode);
+        }
+        return map;
+    }
+
+    private Map<String, ActionMenuHolder> getActionCommands(@NotNull final TreeNode<IMenu> treeNode, @NotNull final List<ActionMenuHolder> parsedHolders)
+    {
+        final Map<String, ActionMenuHolder> map = new HashMap<>();
+        final IMenu menu = treeNode.getData();
+        if (!menu.getMenuType().isNavigationMenu())
+        {
+            final ActionMenu actionMenu = (ActionMenu) menu;
+            for (final ActionArgument actionArgument : actionMenu.getActionArgumentList())
+            {
+                if (usedThisActionArgumentName(parsedHolders, actionArgument))
+                {
+                    continue;
+                }
+                map.put(actionArgument.getName().toLowerCase(Locale.ROOT) + ":", new ActionMenuHolder(treeNode, actionArgument));
+            }
+        }
+        return map;
+    }
+
+    private boolean usedThisActionArgumentName(@NotNull final List<ActionMenuHolder> parsedHolders, @NotNull final ActionArgument actionArgument)
+    {
+        for (final ActionMenuHolder actionMenuHolder : parsedHolders)
+        {
+            if (actionMenuHolder.getActionArgument().getName().equals(actionArgument.getName()))
+            {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    @NotNull
+    @Override
+    public List<String> getTabCompletions(
+                                           @NotNull final MinecraftServer server,
+                                           @NotNull final ICommandSender sender,
+                                           @NotNull final String[] args,
+                                           @Nullable final BlockPos pos)
+    {
+        final ModuleContext moduleContext = getModuleContext();
+        return getTabCompletions(server, sender, args, pos, moduleContext);
+    }
+
+    @NotNull
+    protected List<String> getTabCompletions(
+                                           @NotNull final MinecraftServer server,
+                                           @NotNull final ICommandSender sender,
+                                           @NotNull final String[] args,
+                                           @Nullable final BlockPos pos,
+                                           @NotNull final ModuleContext moduleContext)
+    {
+        try
+        {
+            TreeNode<IMenu> rootTreeNode = getRoot();
+            final ParsingResult parsingResult = getTabCompletionsAndParsingHolders(rootTreeNode, server, sender, args, pos, moduleContext);
+            return parsingResult.getTabCompletions();
+        }
+        catch (CommandException e)
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Either log or rethrow this exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)